### ISSUE TITLE
New Topic: SignalR Configuration

### ISSUE THREAD

**Issue by anurse** [aspnet/Docs#6599]
We should have a docs page that discusses the configuration options available on both the client and the server

* [JSON](https://github.com/aspnet/SignalR/blob/dev/src/Microsoft.AspNetCore.SignalR.Protocols.Json/JsonHubProtocolOptions.cs)/[MessagePack](https://github.com/aspnet/SignalR/blob/dev/src/Microsoft.AspNetCore.SignalR.Protocols.MessagePack/MessagePackHubProtocolOptions.cs) Serialization options (client and server)
* [HubConnection options for configuring the client](https://github.com/aspnet/SignalR/blob/dev/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs#L66)
  * `ServerTimeout` - If the server doesn't respond within this time interval, the connection will be closed
  * `HandshakeTimeout` - If the server doesn't respond to the initial handshake message within this time interval, the connection will be closed
  * `PingInterval` (added in 2.2) - If the client hasn't sent a message within this interval, a Ping message will be sent automatically to keep the connection open
* [HubOptions](https://github.com/aspnet/SignalR/blob/dev/src/Microsoft.AspNetCore.SignalR.Core/HubOptions.cs)
  * `HandshakeTimeout` - If the client doesn't send an initial handshake message within this time interval, the connection will be closed
  * `KeepAliveInterval` - If the server hasn't sent a message within this interval, a Ping message will be sent automatically to keep the connection open
  * `SupportedProtocols` - Protocols supported by this hub (used to disable JSON or MessagePack for certain hubs, for example)
  * `EnableDetailedErrors` - If `true`, sends detailed error messages to the client when exceptions occur (this may contain sensitive data and is `false` by default)

**Comment by BrennanConroy** [comment]
Some more configuration on the server:
* `HubOptions<T>`:
```c#
services.AddSignalR()
    .AddHubOptions<ChatHub>(options =>
    {
        o.EnableDetailedErrors = true;
    }
```

* `HttpConnectionDispatcherOptions`:
  * `ApplicationMaxBufferSize`
  * `AuthorizationData`
  * `LongPolling`
  * `TransportMaxBufferSize`
  * `Transports`
  * `WebSockets`

**Comment by BrennanConroy** [comment]
@anurse Can you look over the details I wrote for the above options. I always get AppMaxBufferSize and TransportMaxBufferSize mixed up

**Comment by anurse** [comment]
I think you are correct.

Some code snippets for `HubConnectionBuilder` configuration:

* Configuring JSON options:

```csharp
var connection = new HubConnectionBuilder()
    .WithUrl("...")
    .AddJsonHubProtocol(options => {
        // options.PayloadSerializerSettings is a JSON.NET JsonSerializerSettings used to 
        // configure how the payload is serialized

        // Example: PascalCase JSON (we are camelCase by default)
        options.PayloadSerializerSettings.ContractResolver = new DefaultContractResolver();        
    })
    .Build();
```

* Restricting the transports that can be used:

```csharp
var connection = new HubConnectionBuilder()
    .WithUrl("...", HttpTransportType.WebSockets | HttpTransportType.ServerSentEvents)
    .Build();
```

* Adding a JWT access token for authentication:

```csharp
var connection = new HubConnectionBuilder()
    .WithUrl("...", options => {
        options.AccessTokenProvider = async () => {
            // Get access token and return it.
        };
    });
    .Build();
```

* Adding headers, cookies and TLS client certificates to requests

```csharp
var connection = new HubConnectionBuilder()
    .WithUrl("...", options => {
        options.Headers["Foo"] = "Bar";
        options.Cookies.Add(new Cookie(...));
        options.ClientCertificates.Add(...);
    });
    .Build();
```

**Comment by BrennanConroy** [comment]
We'll need javascript too

**Comment by ericbrunner** [comment]
@anurse Hi, From your initial issue message all links are dead links. Could you please update them to valid urls. Thx.

- JSON/MessagePack Serialization options (client and server)
- HubConnection options for configuring the client
- HubOptions

**Comment by davidfowl** [comment]
https://docs.microsoft.com/en-us/aspnet/core/signalr/configuration?view=aspnetcore-2.1

**Comment by anurse** [comment]
@ericbrunner Yep, those links were just for reference while writing the docs, and we recently changed our branch structure :). The official doc that @davidfowl is the actual source for information, not this issue (which was just for tracking the work).
